### ISSUE TITLE
Unit testing

### ISSUE THREAD

**Issue by lamarpavel** [lamarpavel/lenie#23]
*lenie* is getting complex enough that sublte bugs are introduced every now and then, so it's valid to think about automated tests. I have not come to any conclusion how *lenie* could be tested; the two choices I see so far are

1. Create a test blog and have a bash script do a number of automated commits
2. Get familiar with a unit testing framework such as [Busted](http://olivinelabs.com/busted/)

I already have a local test blog where I manually push commits after modifying lenie but it's tedious and I doubt that it would be feasable to create automated pushes to cover all corner cases. I am not very familiar with unit testing, so my judgement is limited at the moment.

**Comment by JAEGER2K** [comment]
I already researched this topic and tried to implement unit tests for certain functions. But they all rely too closely to each other. In my opinion we should work to untangle the different tasks/functions and lowering the call stack [1] to make unit testing easier.

Furthermore lenie heavily relies on foreign software (e.g. git), which is executed via `io.popen` or `os.execute`. Also because you can't change the working directory in a lua-script, we'll end up doing both: unittests in lua for lenie and a bash script testing everything else.


----
[1] e.g. `lenie generate foo bar` can result in main() -> generate(arg2,arg3) ->  gen_html() -> gen_posts()

**Comment by lamarpavel** [comment]
Right, I agree with flattening the call stack and untangling all the stuff that happens after one call to `generate()`.

Functions could be grouped into _testable_ and _not testable_ where _testable_ functions are the ones containing the logic and _not testable_ functions are those purely calling other functions. Since the last refactoring of everything under `generate()` and `gen_html()` most of the stuff is already pretty modular, but a closer look at the input parameters reveals that the communication between the functions happens through the content of the tables described in `gen_html()`. This makes testing somewhat harder.

Additionally, the behavious of many functions depends on the global _CONF_ table, which could also be a hurdle for testing.

I'll have a deeper look at Busted because there potentially lots of tests to be carried out and if that can be automated beyond manually writing lots of asserts then I'm happy to invest the time.

**Comment by JAEGER2K** [comment]
*lenie* looks quite stable from my point of view. How about we do some final testing (without unit tests) and if this reveals no bugs we tag it as v1. In the past @Rehwanne has been proven to be a good tester. 

Next step can be some code improvements while retaining full output equality to v1. 
After that we begin to make *lenie* more test friendly and we consider which functions need to be tested and how.
